jlm/llvm/opt/alias-analyses/PointsToGraph.cpp

```diff
@@ -161,6 +161,135 @@ PointsToGraph::AddImportNode(std::unique_ptr<PointsToGraph::ImportNode> node)
+  {
+    if (auto allocaNode = dynamic_cast<const AllocaNode *>(&node))
+    {
+      if (auto it = graph.AllocaNodes_.find(&allocaNode->GetAllocaNode());
+          it != graph.AllocaNodes_.end())
+        return it->second.get();
+    }
+    else if (auto deltaNode = dynamic_cast<const DeltaNode *>(&node))
+    {
+      if (auto it = graph.DeltaNodes_.find(&deltaNode->GetDeltaNode());
+          it != graph.DeltaNodes_.end())
+        return it->second.get();
+    }
+    else if (auto importNode = dynamic_cast<const ImportNode *>(&node))
+    {
+      if (auto it = graph.ImportNodes_.find(&importNode->GetArgument());
+          it != graph.ImportNodes_.end())
+        return it->second.get();
+    }
+    else if (auto lambdaNode = dynamic_cast<const LambdaNode *>(&node))
+    {
+      if (auto it = graph.LambdaNodes_.find(&lambdaNode->GetLambdaNode());
+          it != graph.LambdaNodes_.end())
+        return it->second.get();
+    }
+    else if (auto mallocNode = dynamic_cast<const MallocNode *>(&node))
+    {
+      if (auto it = graph.MallocNodes_.find(&mallocNode->GetMallocNode());
+          it != graph.MallocNodes_.end())
+        return it->second.get();
+    }
+    else if (MemoryNode::Is<UnknownMemoryNode>(node))
+    {
+      return &graph.GetUnknownMemoryNode();
+    }
+    else if (MemoryNode::Is<ExternalMemoryNode>(node))
+    {
+      return &graph.GetExternalMemoryNode();
+    }
+    else
+      JLM_UNREACHABLE("Unknown type of MemoryNode");
+
+    return nullptr;
+  };
+
+  // Given two nodes, checks if the first node points to everything the second points to.
+  // The nodes can belong to different PointsToGraphs.
+  const auto HasSupersetOfPointees = [&GetCorrespondingMemoryNode](
+                                         const PointsToGraph::Node & superset,
+                                         const PointsToGraph::Node & subset)
+  {
+    for (const auto & subsetTarget : subset.Targets())
+    {
+      auto correspondingTarget = GetCorrespondingMemoryNode(subsetTarget, superset.Graph());
+
+      // Check if the superset is pointing to the target
+      if (correspondingTarget == nullptr || !superset.HasTarget(*correspondingTarget))
+        return false;
+    }
+    return true;
+  };
+
+  // Given a memory node from the subgraph, check that a corresponding node exists in this graph.
+  // All edges the other node has, must have corresponding edges in this graph.
+  // If the other node is marked as escaping, the corresponding node must be as well.
+  const auto HasSuperOfMemoryNode = [&](const PointsToGraph::MemoryNode & subNode)
+  {
+    const auto thisNode = GetCorrespondingMemoryNode(subNode, *this);
+    if (thisNode == nullptr)
+      return false;
+
+    if (subNode.IsModuleEscaping() && !thisNode->IsModuleEscaping())
+      return false;
+
+    return HasSupersetOfPointees(*thisNode, subNode);
+  };
+
+  // Iterate through all memory nodes in the subgraph, and make sure we have corresponding nodes
```

Comment:
Maybe compare the number of various nodes between the two graphs first before you iterate through everything?